### PR TITLE
Added the possibility to customise events with datadog metadata

### PR DESCRIPTION
**Pull request by ajacquemot**
### What does this PR do?

* Added the possibility to enrich log events with a service and some tags, settable at the sink configuration.
* Added a source: `csharp`

### Motivation

Enrich log events and fully benefit from Datadog features.

### Additional Notes

Still need to update the REAME.md

### CODE REVIEW

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogMessage.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            Event = logEvent;
+            Source = DDSource;
+            Service = service;
+            Tags = tags;
+        }
+
+        /// <summary>
+        /// Returns the JSON representation of the message.
+        /// </summary>
+        public override string ToString()
```
Comment (bmermet) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogMessage.cs]: I'm not sure if it makes that much of a difference, but I've never seen the `ToString` function used for serialization. It's usually meant as a short summary of your object that can be printed by the debugger and serialization is done by calling `JsonConvert.SerializeObject` on the object itself, which makes the serialization format explicit

Also having a dependency on `Newtonsoft.Json` in `ToString` could mess with the debugger if there was an issue loading the `Newtonsoft.Json` dependency.

**Comment by ian28223** [comment]
Can we set the `source` to `serilog` or (ideally) `messagetemplates`  by default if the user has not set any (intentionally or not)? Why `messagetemplates` and not `csharp`? This would allow us potentially create an integration pipeline for it in the future should more languages/libraries adopt  this messagetemplates format. + serilog is just one of [many implementations](https://messagetemplates.org/#implementations) supporting this format; e.g. python, ruby, java

File: src/Serilog.Sinks.Datadog.Logs/LoggerConfigurationDatadogLogsExtensions.cs [hunk 2]
```diff
@@ -20,12 +20,18 @@ public static class LoggerConfigurationDatadogExtensions
         /// </summary>
         /// <param name="loggerConfiguration">The logger configuration.</param>
         /// <param name="apiKey">Your Datadog API key.</param>
+        /// <param name="source">The integration name.</param>
+        /// <param name="service">The service name.</param>
+        /// <param name="tags">Custom tags.</param>
         /// <param name="configuration">The Datadog logs client configuration.</param>
         /// <returns>Logger configuration</returns>
         /// <exception cref="ArgumentNullException">A required parameter is null.</exception>
         public static LoggerConfiguration DatadogLogs(
```
Comment (rramer-web-com) [inline src/Serilog.Sinks.Datadog.Logs/LoggerConfigurationDatadogLogsExtensions.cs]: I'd like to see "host" as a settable attribute here as well that flows into a DataDogMessage.Host property.

**Comment by stevejgordon** [comment]
This looks like exactly the functionality I was missing when testing this sink today. Is there an ETA for this to be merged and a release to be made on NuGet containing the changes?

**Comment by dreznicek** [comment]
Any update on this? Its been sitting for 2 months.

**Comment by NBParis** [comment]
Hello everyone,

Sorry for the delay on this feature.
Until this is released, there are a couple workarounds that can be used today as explained in [this issue](https://github.com/DataDog/serilog-sinks-datadog-logs/issues/7).

That said, I'll make sure to raise this internally so it can be released as soon as possible.

**Comment by mehalick** [comment]
Just stumbled on this issue for exactly the same thing, having the ability to enrich logs with host and tag information would be extremely useful.

File: README.md [hunk 3]
```diff
@@ -11,20 +11,28 @@ var log = new LoggerConfiguration()
...
+```csharp
 var config = new DatadogConfiguration("intake.logs.datadoghq.com", 10516, true);
 var log = new LoggerConfiguration()
-    .WriteTo.DatadogLogs("<API_KEY>", config)
+    .WriteTo.DatadogLogs(
+        "<API_KEY>",
+        source: "<SOURCE_NAME>",
+        service: "<SERVICE_NAME>",
+        host: "<HOST_NAME>",
+        tags: new string[] {"<TAG_1:VALUE_1", "TAG_2:VALUE_2>"},
```
Comment (ajacquemot) [inline README.md]: nits: `{"<TAG_1:VALUE_1", "TAG_2:VALUE_2>"}` -> `{"<TAG_1>:<VALUE_1>", "<TAG_2>:<VALUE_2>"}`

File: README.md [hunk 4]
```diff
@@ -11,20 +11,28 @@ var log = new LoggerConfiguration()
     .CreateLogger();
 ```
 
-You can override the default behavior by manually specifing the following properties (endpoint, port, useSSL).
+You can override the default behavior by manually specifing the following properties (endpoint, port, useSSL). You can also add
```
Comment (ajacquemot) [inline README.md]: nits, I'd move `You can also add ..` to a new line

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogMessage.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            Service = service;
+            Host = host;
+            Tags = tags;
+
+            if (logEvent.Properties != null) {
+                Properties = new Dictionary<string, object>();
+                foreach (var key in logEvent.Properties.Keys)
+                {
+                    var value = logEvent.Properties[key];
+                    Properties.Add(key, value.ToString());
```
Comment (ajacquemot) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogMessage.cs]: We should not transform the value into a String here and keep an object instead that.

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs [hunk 6]
```diff
@@ -53,6 +65,25 @@ public DatadogSink(string apiKey, string source, string service, string host, st
             _client = new DatadogClient(config);
         }
 
+        private string formatMessage(LogEvent logEvent, string source, string service, string host, string tags) {
```
Comment (ajacquemot) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs]: Can we add a comment to the method please to explain why we need to do this ?

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs [hunk 7]
```diff
@@ -53,6 +65,25 @@ public DatadogSink(string apiKey, string source, string service, string host, st
...
+            var payload = new StringBuilder();
+            var writer = new StringWriter(payload);
+
+            // Serialize the event as JSON. The Serilog formatter handles the
+            // internal structure of the logEvent to give a nicely formatted JSON
+            formatter.Format(logEvent, writer);
+
+            // Convert the JSON to a dictionnary and add the DataDog properties
+            var logEventAsDict = JsonConvert.DeserializeObject<Dictionary<string, dynamic>>(payload.ToString());
+            logEventAsDict.Add("ddsource", source);
```
Comment (ajacquemot) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs]: What happens if the entry already exists ? What should we do ?

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs [hunk 8]
```diff
@@ -53,6 +65,25 @@ public DatadogSink(string apiKey, string source, string service, string host, st
...
+            logEventAsDict.Add("service", service);
+            logEventAsDict.Add("host", host);
+            logEventAsDict.Add("ddtags", tags);
+
+            // Convert back the dict to a JSON
```
Comment (ajacquemot) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs]: nits: `to a JSON string`

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs (hunk 7 above)
Comment (achntrl) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs]: The object we serialize and deserialize is a LogEvent. The keys that this object have are controlled: https://github.com/serilog/serilog/blob/9a46e68d3e224b965f4ac2665fd379ae9568d6b4/src/Serilog/Formatting/Json/JsonFormatter.cs#L117-L157
They also are capitalized, so we don't risk any collision

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs [hunk 9]
```diff
@@ -75,12 +78,13 @@ public DatadogSink(string apiKey, string source, string service, string host, st
 
             // Convert the JSON to a dictionnary and add the DataDog properties
             var logEventAsDict = JsonConvert.DeserializeObject<Dictionary<string, dynamic>>(payload.ToString());
-            logEventAsDict.Add("ddsource", source);
-            logEventAsDict.Add("service", service);
-            logEventAsDict.Add("host", host);
-            logEventAsDict.Add("ddtags", tags);
+            if ( source!= null) { logEventAsDict.Add("ddsource", source); }
```
Comment (ajacquemot) [inline src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs]: nits: `source!= null` -> `source != null` and same for all the following parameters

File: src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs (hunk 7 above)
Comment (ajacquemot) [reply src/Serilog.Sinks.Datadog.Logs/Sinks/Datadog/DatadogSink.cs]: 👍

**Pull request merged**